### DISCUSSION TITLE
Possibility to update PUSH_INSTALLATION_* variables in the admin panel

### DISCUSSION THREAD

**Discussion by Getslow6** [dani-garcia/vaultwarden#4164] [Ideas]
Hi,

Maybe this has been requested before already, but couldn't find it. 

As a Home Assistant user, I'm running the Vaultwarden container as an addon to Home Assistant. In the Home Assistant Operating System we don't have direct access to the environment variables, as the Vaultwarden container is 'wrapped' to become the Home Assistant addon.

In order to enable the push notification service, we need to be able to set the following environment variables:
```
PUSH_ENABLED=true
PUSH_INSTALLATION_ID=CHANGEME
PUSH_INSTALLATION_KEY=CHANGEME
```

Two feature requests have been opened in he [Vaultwarden Addon](https://github.com/hassio-addons/addon-bitwarden) project:
- https://github.com/hassio-addons/addon-bitwarden/issues/283
- https://github.com/hassio-addons/addon-bitwarden/issues/278

The maintainer mentioned that this should preferably be added upstream - here - which makes sense in my opinion.

Would it be possible to make the environment variables adaptable from the Admin page?

**Comment by BlackDex** [comment]
Those are env variables already which can be used.
Since these are read only they can't be changed in the admin panel, like many others, like `DATABASE_URL` for example.

You should just somehow add those as env variables and it should work.

These variables are read only for a reason.
The only other way i can offer you is to manually modify the `config.json` in your data folder and restart.

With a note that we normally do not recommend to edit this file manually.

**Reply by Getslow6** [reply]
Thank you for your reply.

For my information; what is the reason for these to be read only (can't be set in runtime?) ?

If due to these reasons these variables will never be editable from the admin panel, it might be a reason to add a workaround in the Home Assistant addon downstream.

**Reply by BlackDex** [reply]
Some variables are only read once during startup, or are lazy loaded to prevent a lot of overhead extracting those values everytime it is needed, mostly for performance.

Setting those via the admin panel indicates a direct active setting. To prevent confusion and issues, we only allow those to be set via env variables.

Maybe it the future we can make some changes if we can trigger a full reload of the app without a container going down and other issues. But currently that is not an option.

**Reply by clementko** [reply]
I am a Home Assistant Vaultwarden Addon user. After looking at forking and add the push enabled flag, id and key myself in the HA addon, I settled with editing the /data/config.json inside the docker directly. so far it is working fine
